scheduler/scheduler-examples/src/main/java/org/ow2/proactive/scheduler/examples/SFTPConnector.java

```diff
@@ -0,0 +1,366 @@
+                    // Create remote directory(ies) if it(they) does(do) not exist
+                    if (!remoteDirExists(channelsftp, sftpRemoteRelativePath)) {
+                        makeRemoteDirectories(channelsftp, Paths.get(sftpRemoteRelativePath));
+                    }
+                    // recursive folder content upload to sftp server
+                    filesRelativePathName = recursiveFolderUpload(channelsftp,
+                                                                  sftpLocalRelativePath,
+                                                                  sftpRemoteRelativePath);
+                    getOut().println("END UPLOAD TO SFTP");
+                    break;
+
+                default:
+                    throw new IllegalArgumentException("SFTP mode can only be PUT or GET.");
+            }
+        } catch (JSchException e) {
+            throw new JSchException("An error occured while trying to connect to SFTP server: ", e);
+        } catch (SftpException e) {
+            throw new SftpException(e.id,
+                                    "File not found, permission denied or operation not supported by SFTP server.",
+                                    e);
+        } finally {
+            if (channelsftp != null)
+                channelsftp.disconnect();
+            if (channel != null)
+                channel.disconnect();
+            if (session != null)
+                session.disconnect();
+            getOut().println("Disconnected from SFTP server");
+        }
+
+        return (Serializable) filesRelativePathName;
+    }
+
+    /**
+     * This method recursively uploads a local file/folder to an SFTP server
+     *
+     * @param sourcePath
+     * @param destinationPath
+     * @throws SftpException
+     * @throws FileNotFoundException
+     */
+    private List<String> recursiveFolderUpload(ChannelSftp channelSftp, String sourcePath, String destinationPath)
+            throws SftpException, FileNotFoundException {
+
+        List<String> filesRelativePathName = new ArrayList<>();
+
+        File sourceFile = new File(sourcePath);
+        if (sourceFile.isFile()) {
+            // copy if it is a file
+            channelSftp.cd(destinationPath);
+            getOut().println("Uploading " + sourcePath);
+            channelSftp.put(new FileInputStream(sourceFile), sourceFile.getName(), ChannelSftp.OVERWRITE);
+            getOut().println("File uploaded successfully to " +
+                             Paths.get(destinationPath, sourceFile.getName()).toString());
+            filesRelativePathName.add(Paths.get(destinationPath, sourceFile.getName()).toString());
+
+        } else {
+
+            String remoteFilePath = Paths.get(destinationPath, sourceFile.getName()).toString();
+            File[] files = sourceFile.listFiles();
+
+            if (files != null && !sourceFile.getName().startsWith(".")) {
+
+                channelSftp.cd(destinationPath);
```

Comment:
this looks very strange, and looks like a bug, unless destinationPath is always absolute ! As the channelSftp is passed recursively, it has a memory of the remote current folder.

What happens when you want to transfer:
a/b/file_c
d/file_e
?

To me, it will do:
mkdir a, cd a, mkdir b, cd b, transfer file_c
mkdir d, cd d,  transfer file_e
And this will result in the structure:
a/b/file_c
a/b/d/file_e

instead of what is needed.

Maybe this code works, it is hard to tell, because the distinction relative/absolute is not clear enough

Reply:
In fact, it seems a bit confusing, but it works perfectly either with relative or absolute paths (see the attached screenshot).

![folder-structure](https://user-images.githubusercontent.com/1918419/38024986-0b96f57e-3287-11e8-90c1-04f8c58147ce.png)